### ISSUE TITLE
async validation

### ISSUE THREAD

**Issue by sysmat** [validointi/validointi#87]
- is it possible to have vest async validation?
- I try but no luck
- dependencies
```
"@validointi/core": "^1.4.2",
"vest": "^5.2.12",
"@angular/core": "~17.3.6",
``` 
- code:

```typescript
const createSimpleFormValidations = (infoService: InfoService) => {
  return create((data: Model = {} as Model) => {
    validateName(data);
    omitWhen(!data.id, () => {
      test('id', 'User id is already taken', async ({ signal }) => {
        return lastValueFrom(infoService.userIdExists(data.id!).pipe(takeUntil(fromEvent(signal, 'abort')))).then(
          () => {
            console.log('reject');
            return Promise.reject();
          },
          () => {
            console.log('resolve');
            return Promise.resolve();
          }
        );
      });
    });
  });
};

// component
#vr = inject(ValidatorRegistryService);
#infoService = inject(InfoService);

validate = this.#vr.registerValidator(
    'sample-data',
    createVestAdapter(createSimpleFormValidations(this.#infoService))
  );
model: Model = { name: 'This can be your own name', id: null };
```

```html
<form
      validationId="sample-data"
      (ngSubmit)="onSubmit($any(form))"
      #form="ngForm">
      <label for="name">
        <span>Your name</span>
      </label>
      <input
        type="text"
        name="name"
        placeholder="Fill in your name"
        [(ngModel)]="model.name"
        #name="ngModel" />

      @if (name.errors) {
        <p>{{ name.errors['name'] }}</p>
      }
      <pre>{{ name.errors | json }}</pre>
      <br />
      <label for="id">
        <span>Id</span>
      </label>
      <input
        type="number"
        name="id"
        placeholder="Fill in your id"
        [(ngModel)]="model.id"
        #id="ngModel" />

      @if (id.errors) {
        <p>{{ id.errors['id'] }}</p>
      }
      <pre>{{ id.errors | json }}</pre>
  
      <br />
      <input
        type="submit"
        value="Submit"
        [disabled]="form.invalid" />
    </form>
```
- in network tab I see request/response and reject, but  `{{ id.errors | json }}` is always null

**Comment by SanderElias** [comment]
the `lastValueFrom` wait for the complete if the inner observable, and from the looks of it, it doesn't complete.

The Validointi library is entirely suited to handle async validation. (all things are designed to be async from the get-go)

**Comment by sysmat** [comment]
It complete, just not working, do you have any example for async validation?

**Comment by Jefiozie** [comment]
We have a async example here

[Async](https://github.com/validointi/validointi/blob/3a47c4edb2b4ae07312c3c1012662265ad9ac0e8/projects/demo/src/app/form2/sample-data-form2.service.ts#L65)

**Comment by sysmat** [comment]
- I switch to fetch, and the same not working, there must be something wrong in my app, I can't figure it out
- @Jefiozie thx for example

**Comment by SanderElias** [comment]
If your app is available on GH, I can take a quick look?

**Comment by sysmat** [comment]
I also have typescript problems with `test('name', 'name is required', () => checkValueOnEndpoint(data.name));`

```
Overload 1 of 4, '(fieldName: "id", message: string, cb: TestFn): TIsolateTest<string, string>', gave the following error.
    Type 'Promise<string>' is not assignable to type 'TestResult'.
      Type 'Promise<string>' is not assignable to type 'AsyncTest'.
        Type 'string' is not assignable to type 'void'.
  Overload 2 of 4, '(fieldName: "id", cb: TestFn, key: IsolateKey): TIsolateTest<string, string>', gave the following error.
    Argument of type 'string' is not assignable to parameter of type 'TestFn'. [plugin angular-compiler]

    src/app/vest-form-validation.component.ts:29:6:
      29 │       test('id', 'User id is already taken', () => infoService.use...
         ╵       ~~~~

  The expected type comes from the return type of this signature.

    node_modules/vest/types/vest.d.ts:27:14:
      27 │ type TestFn = (payload: TestFnPayload) => TestResult;
         ╵               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by SanderElias** [comment]
I think your test is async, but your suite isn't 
From the Validointi side, the suite is called from the adapter like this:

```typescript
export function createVestAdapter<T>(suite: VestLike<T>) {
  return async (data: T, field?: string) => {
    const errors = (await suite(data, field).getErrors()) as Record<string, string[]>;
    return Object.entries(errors).reduce((acc, [key, err]) => ({ ...acc, [key]: err }), {} as ValidationErrors);
  };
}
```
As you can see, it is fully prepared to handle async (it is even async by default)
Your type errors also seem to relate to mixing sync and async in your suite.

**Comment by sysmat** [comment]
Is this example working in repo, very strange

**Comment by sysmat** [comment]
- in example checkValueOnEndpointis not returning  this types :
```
type TestFn = (payload: TestFnPayload) => TestResult;
type AsyncTest = Promise<void>;
type TestResult = Maybe<AsyncTest | boolean> | void;
```
and ts is ok with it

**Comment by sysmat** [comment]
hm, also example not working in my app, typescript problems with ¸

```
X [ERROR] TS2345: Argument of type 'Promise<ValidationErrors>' is not assignable to parameter of type 'Validator<Object>'.
  Type 'Promise<ValidationErrors>' provides no match for the signature '(data: Object, fieldName?: keyof Object | undefined, group?: string | undefined): Promise<ValidationErrors>'. [plugin angular-compiler]

    src/app/vest-form-validation.component.ts:108:55:
      108 │ ...le-data', createSimpleFormValidations(this.#infoService, this....
          ╵              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

- hm, I use vest in svelte(typescript), but here in angular I'll do it all fasion with async validators from reactive forms, it is to many problems with types

**Comment by sysmat** [comment]
example is a bug

```
X [ERROR] TS2769: No overload matches this call.
  Overload 1 of 4, '(fieldName: "name", message: string, cb: TestFn): TIsolateTest<string, string>', gave the following error.
    Type 'Promise<string>' is not assignable to type 'TestResult'.
      Type 'Promise<string>' is not assignable to type 'AsyncTest'.
        Type 'string' is not assignable to type 'void'.
  Overload 2 of 4, '(fieldName: "name", cb: TestFn, key: IsolateKey): TIsolateTest<string, string>', gave the following error.
    Argument of type 'string' is not assignable to parameter of type 'TestFn'. [plugin angular-compiler]

    src/app/service/sample-data-form2.service.ts:45:4:
      45 │     test('name', 'name is required', () => checkValueOnEndpoint(da...
         ╵     ~~~~

  The expected type comes from the return type of this signature.

    node_modules/vest/types/vest.d.ts:27:14:
      27 │ type TestFn = (payload: TestFnPayload) => TestResult;
         ╵               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by sysmat** [comment]
yours stackblitz is using very old versions of everything

**Comment by sysmat** [comment]
- I create fork, update dependencies, and using [async](https://github.com/validointi/validointi/blob/3a47c4edb2b4ae07312c3c1012662265ad9ac0e8/projects/demo/src/app/form2/sample-data-form2.service.ts#L65)
- fork on stackblitz [bug-report](https://stackblitz.com/edit/validointi-sample-qlo9uf?file=src%2Fapp%2Ftemplate-form%2Ftemplate-form.component.ts)

**Comment by SanderElias** [comment]
@Jefiozie I don't have time to test this with Vest 4 right now, would you be so kind to take a look?

**Comment by Jefiozie** [comment]
Based on what I've tested everything still works.

**Comment by Jefiozie** [comment]
@sysmat I've created a issue where we will update our examples

**Comment by sysmat** [comment]
@Jefiozie with updated version it still work, strange I have typescript errors

**Comment by SanderElias** [comment]
@sysmat So, your validation works, but you have TypeScript errors?
What errors do you have? Do you still have the same as above? 
It will be easier to help you if you have a reproduction of your errors in a stack blitz.

**Comment by sysmat** [comment]
@SanderElias no it can not work because ng compile fail, you can se in bug-report

**Comment by SanderElias** [comment]
@sysmat  I can't reproduce your errors. 
If you can not provide a repo or a stack blitz with a reproduction of your issue, there is nothing actionable left in this issue.

**Comment by sysmat** [comment]
- are looking at https://stackblitz.com/edit/validointi-sample-qlo9uf?file=src%2Fapp%2Ftemplate-form%2Ftemplate-form.component.ts
- when type your name: you will se only 2 sync errors, but not async, if you open browser console you will see that async validator is called but not shown in UI

**Comment by SanderElias** [comment]
I added it to your [StackBlitz](https://stackblitz.com/edit/validointi-sample-3d9x4g?file=src%2Fapp%2Ftemplate-form%2Ftemplate-form.component.ts):
![image](https://github.com/validointi/validointi/assets/1249083/e0572197-3160-4da8-b831-03de410bfcd0)

**Comment by sysmat** [comment]
- hm, I get browser errors

```
ERROR TypeError: suite(...).getErrors is not a function
    at validointi-core.mjs:492:45
    at Generator.next (<anonymous>)
    at chunk-5ERGMWRA.js?v=3430041a:101:61
    at new ZoneAwarePromise (zone.js:1425:21)
    at __async (chunk-5ERGMWRA.js?v=3430041a:85:10)
    at validointi-core.mjs:491:33
    at _ValidatorDirective.<anonymous> (validointi-core.mjs:221:28)
    at Generator.next (<anonymous>)
    at fulfilled (chunk-5ERGMWRA.js?v=3430041a:88:24)
    at _ZoneDelegate.invoke (zone.js:368:26)
```

- and not working in UI, did you change anything else(haha, ts as any)

**Comment by sysmat** [comment]
aha, I see you update "@validointi/core": "^1.6.0",

**Comment by SanderElias** [comment]
@sysmat  Yes, there was a breaking change, as I said above, between Vest 4 and 5. I updated Validointi to deal with this.
I removed the `as any` from the stackblitz as it was not needed. I assumed you had updated the stackblitz to use the latest version, and I was wondering why I would get an error. I assumed wrongly. Everything worked as expected when I did the actual update. 
Forgot to take out the any.
